### PR TITLE
add Kokkos

### PR DESCRIPTION
**Pull request by Crghilardi**
This PR adds a `build_tarballs.jl` for the [Kokkos](https://github.com/kokkos/kokkos) library.This build is close to default build options, only added openMP support.

Tested locally on `x86_64-linux-*` , `powerpc`, and `freebsd`. 
`mingw` still has some issues I am working through

### CODE REVIEW

**Comment by Crghilardi** [comment]
As far as I understand, `mingw` is failing with errors about the use of `_InterlockedExchangeCompare8`
```
[01:28:52] cd /workspace/srcdir/kokkos-3.4.01/build/core/src && /opt/bin/x86_64-w64-mingw32-libgfortran3-cxx11/x86_64-w64-mingw32-g++ --sysroot=/opt/x86_64-w64-mingw32/x86_64-w64-mingw32/sys-root/  -DKOKKOS_DEPENDENCE -Dkokkoscore_EXPORTS @CMakeFiles/kokkoscore.dir/includes_CXX.rsp -O3 -DNDEBUG   -o CMakeFiles/kokkoscore.dir/impl/Kokkos_hwloc.cpp.obj -c /workspace/srcdir/kokkos-3.4.01/core/src/impl/Kokkos_hwloc.cpp
[01:28:54] In file included from /workspace/srcdir/kokkos-3.4.01/core/src/Kokkos_Atomic.hpp:187:0,
[01:28:54]                  from /workspace/srcdir/kokkos-3.4.01/core/src/impl/Kokkos_HostBarrier.hpp:49,
[01:28:54]                  from /workspace/srcdir/kokkos-3.4.01/core/src/impl/Kokkos_HostBarrier.cpp:47:
[01:28:54] /workspace/srcdir/kokkos-3.4.01/core/src/impl/Kokkos_Atomic_Windows.hpp: In function ‘T Kokkos::atomic_compare_exchange(volatile T*, const T&, typename std::enable_if<(sizeof (T) == sizeof (char)), const T&>::type)’:
[01:28:54] /workspace/srcdir/kokkos-3.4.01/core/src/impl/Kokkos_Atomic_Windows.hpp:88:58: error: there are no arguments to ‘_InterlockedCompareExchange8’ that depend on a template parameter, so a declaration of ‘_InterlockedCompareExchange8’ must be available [-fpermissive]
[01:28:54]                                         *((CHAR*)&compare));
[01:28:54]                                                           ^
[01:28:54] /workspace/srcdir/kokkos-3.4.01/core/src/impl/Kokkos_Atomic_Windows.hpp:88:58: note: (if you use ‘-fpermissive’, G++ will accept your code, but allowing the use of an undeclared name is deprecated)
[01:28:54] make[2]: *** [core/src/CMakeFiles/kokkoscore.dir/build.make:140: core/src/CMakeFiles/kokkoscore.dir/impl/Kokkos_HostBarrier.cpp.obj] Error 1
```
This seems to be defined as a [MSVC intrinsic](https://docs.microsoft.com/en-us/cpp/intrinsics/interlockedcompareexchange-intrinsic-functions?view=msvc-160), but is missing from `mingw`. It looks like there are [64 and 128 versions ](https://sourceforge.net/p/mingw-w64/code/HEAD/tree/trunk/mingw-w64-headers/crt/intrin.h#l353) in `intrin.h` but no 8 version.

I tried adding both `#include <intrin.h>` and `#include <winnt.h>` statement as a wild guess, but that still returns the same error.

File: K/Kokkos/build_tarballs.jl [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+"${OPENMP_FLAG[@]}"
+
+make -j${nproc}
+make install
+
+"""
+
+# These are the platforms we will build for by default, unless further
+# platforms are passed in on the command line
+platforms = expand_cxxstring_abis(supported_platforms())
```
Comment (giordano) [inline K/Kokkos/build_tarballs.jl:51]:
```suggestion
platforms = expand_cxxstring_abis(supported_platforms(; experimental=true))
```

File: K/Kokkos/build_tarballs.jl [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+
+
+# The products that we will ensure are always built
+products = [
+    LibraryProduct("libkokkoscore", :libkokkoscore),
+    LibraryProduct("libkokkoscontainers", :libkokkoscontainers)
+]
+
+# Dependencies that must be installed before this package can be built
+dependencies = Dependency[
```
Comment (giordano) [inline K/Kokkos/build_tarballs.jl:63]:
```suggestion
dependencies = [
```

**Comment by giordano** [comment]
I'm a bit confused: what was that made the MinGW build work?

**Comment by Crghilardi** [comment]
I don't know exactly, there were [a lot of changes and files modified](https://github.com/kokkos/kokkos/compare/3.4.01...3.5.00) between the 3.4.1 release and 3.5.0. I am just guessing some macro flags or something was mis-aligned or not triggering?

 I don't know if my ramblings above were actually meaningful in any way, so take them with a grain of salt

**Comment by giordano** [comment]
Ok, I missed you had upgraded the version in the meantime, that may have had an effect then.

**Comment by Crghilardi** [comment]
I also attempted to submit a patch upstream [here](https://github.com/kokkos/kokkos/pull/4564) for completeness.

**Pull request merged**